service/history/workflow/mutable_state_impl.go

```diff
@@ -4236,6 +4236,51 @@ func (ms *MutableStateImpl) RejectWorkflowExecutionUpdate(_ string, _ *updatepb.
 	return nil
 }
 
+func (ms *MutableStateImpl) AddWorkflowExecutionOptionsUpdatedEvent(
+	versioningOverride *workflowpb.VersioningOverride,
+) (*historypb.HistoryEvent, error) {
+	if err := ms.checkMutability(tag.WorkflowActionWorkflowOptionsUpdated); err != nil {
+		return nil, err
+	}
+	event := ms.hBuilder.AddWorkflowExecutionOptionsUpdatedEvent(versioningOverride)
+	if err := ms.ApplyWorkflowExecutionOptionsUpdatedEvent(event); err != nil {
+		return nil, err
+	}
+	return event, nil
+}
+
+func (ms *MutableStateImpl) ApplyWorkflowExecutionOptionsUpdatedEvent(event *historypb.HistoryEvent) error {
+	override := event.GetWorkflowExecutionOptionsUpdatedEventAttributes().GetVersioningOverride()
+	previousCurrentDeployment := ms.GetEffectiveDeployment()
+	previousEffectiveVersioningBehavior := ms.GetEffectiveVersioningBehavior()
+	ms.GetExecutionInfo().GetVersioningInfo().GetBehavior()
+	ms.GetExecutionInfo().VersioningInfo.VersioningOverride = override
+	if !proto.Equal(ms.GetEffectiveDeployment(), previousCurrentDeployment) ||
+		// If behavior changes from pinned --> unpinned with the same effective deployment, we still need to reschedule
```

Comment:
I think this is also needed for unpinned --> pinned direction because the matching queue would change even if the deployment stays the same. (logic seems ok but the comment doesn't have to be for one direction only)